src/streaming_partial_decompresser.js

```diff
@@ -0,0 +1,160 @@
+const fs = require('fs');
+const zstd = require('node-zstandard');
+const exec = require('child_process').exec;
+const { spawn } = require('child_process');
+
+// List of all the database files you want to download
+const files = ["lichess_db_standard_rated_2013-01.pgn.zst", /*...*/];
+
+// 30 games = 10*1024 bytes, 1 game = 350 bytes, 1000 games = 330KB, 100K games = 33MB
+// 10MB yields around 30k games, 5GB = 15 million games?
+const SIZE_LIMIT = 10 * 1024 * 1024
+
+// function to check file size
+const getFileSize = (filePath) => {
+    if (!fs.existsSync(filePath)) {
+        return 0;
+    }
+    const stats = fs.statSync(filePath);
+    return stats.size;
+};
+
+const runAnalysis = (filePath) => {
+    return new Promise((resolve, reject) => {
+        // Run the analysis script
+        console.log(`Running analysis script on ${filePath}...`);
+
+        const child = spawn('ts-node', ['/Users/bennyrubanov/chessanalysis/src/index.ts', filePath]);
+
+        child.stdout.on('data', (data) => {
+            console.log(`stdout: ${data}`);
+        });
+
+        child.stderr.on('data', (data) => {
+            console.log(`stderr: ${data}`);
+        });
+
+        child.on('error', (error) => {
+            console.log(`error: ${error.message}`);
+            reject(error);
+        });
+
+        child.on('close', (code) => {
+            console.log(`child process exited with code ${code}`);
+            resolve();
+        });
+    });
+};
+
+const decompressAndAnalyze = async (file, start = 0) => {
+    let chunk_counter = 0; // Initialize the chunk counter
+    let file_counter = 0; // Initialize the file counter
+
+    const base_path = `/Users/bennyrubanov/chessanalysis/data/${file.replace('.zst', '')}`;
+
+    // Create a new file path
+    let path = `${base_path}_${file_counter}`;
+
+    // Create a new writable stream
+    let decompressedStream = fs.createWriteStream(path, { flags: 'a' });
+
+    // Check if file already exists
+    if (fs.existsSync(path)) {
+        const stats = fs.statSync(path);
+        start = stats.size;
+    }
+
+    try {
+        await new Promise((resolve, reject) => {
+            console.log(`Starting decompression of chunk number ${chunk_counter}.`);
+
+            let startTime = Date.now();
+
+            zstd.decompressionStreamFromFile(`/Users/bennyrubanov/chessanalysis/data/${file}`, (err, result) => {
+                if (err) return reject(err);
+
+                let lastChunkLength = 0;
+                let fileLength = 0;
+
+                result.on('error', (err) => {
+                    return reject(err);
+                });
+
+                result.on('data', async (data) => {
+                    decompressedStream.write(data);
```

Comment:
do you clear the stream? Are you overwriting it? Won't this end up being OOM?

Reply:
decompressedStream.write(data) appends the data to the file, not storing it in memory. shouldn't cause an out-of-memory (OOM) error

the strat is to not stop the stream, write it to a file until the file hits the size limit, then immediately redirect the stream to a new file and start analysis on the old one, rinse and repeat

Reply:
so data is continuously being written to disk, not stored in memory. doesn't clear stream 

also trying to implement old file deletion once analysis is done

basically parallel processes: stream write to new file, analysis + deletion of old file